node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1294,7 +1335,135 @@ async fn fetch_block_worker(
+    fn new() -> Self {
+        Self(None)
+    }
+
+    fn add_signatures(&mut self, maybe_signatures: Option<BlockSignatures>) {
+        let signatures = match maybe_signatures {
+            Some(sigs) => sigs,
+            None => return,
+        };
+
+        match &mut self.0 {
+            None => {
+                self.0 = Some(signatures);
+            }
+            Some(old_sigs) => {
+                for (pub_key, signature) in signatures.proofs {
+                    old_sigs.insert_proof(pub_key, signature);
+                }
+            }
+        }
+    }
+
+    fn check_if_sufficient(
+        &self,
+        validator_weights: &BTreeMap<PublicKey, U512>,
+        finality_threshold_fraction: Ratio<u64>,
+    ) -> bool {
+        self.0.as_ref().map_or(false, |sigs| {
+            consensus::check_sufficient_finality_signatures(
+                validator_weights,
+                finality_threshold_fraction,
+                sigs,
+            )
+            .is_ok()
+        })
+    }
+
+    fn into_inner(self) -> Option<BlockSignatures> {
+        self.0
+    }
+}
+
+// Fetches the finality signatures from the given peer. In case of timeout, it'll
+// retry up to `retries` times. Other errors interrupt the process immediately.
+async fn fetch_finality_signatures_with_retry(
+    block_hash: BlockHash,
+    peer: NodeId,
+    retries: usize,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<FetchedData<BlockSignatures>, FetcherError<BlockSignatures>> {
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_hash(
+    block_hash: BlockHash,
+    validator_weights: &BTreeMap<PublicKey, U512>,
+    ctx: &ChainSyncContext<'_>,
+) -> Option<BlockSignatures> {
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+        let maybe_signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_hash,
+                    ?peer,
+                    "did not get FinalitySignatures from peer, got from storage instead",
+                );
+                Some(*item)
```

Comment:
Is `BlockSignature` the item type that returns them from storage even if they aren't sufficient (in that case we would fail to ask a peer here) or the one that doesn't (in that case we could just return there because we already have them)?

Reply:
Note that this one might still be a blocker. We need to make sure that:

* The requesting node will ask for signatures from a peer as long as it doesn't have a quorum of them in storage, even if it has some in storage.
* The responding node will respond with signatures from its storage even if it has fewer than a quorum.

I believe that means the requesting and responding node can't use the same `Item` implementation for it, or the `Item` implementation needs to have some special logic for this.

Reply:
No, a single `Item` is perfectly fine, actually. The code doesn't actually live in `Item`, BTW - there is a part in the implementation of `ItemFetcher<T>`, and a part is in the `Storage` component.

Now this will work perfectly fine, because `Fetcher<BlockSignatures>` will only return the result from storage if the signatures are sufficient, but if they are not, it will ask other peers, and - here's the important part - the requests won't go through the fetchers on the peers. `GetRequest`s are handled directly by storage, and the code there will return a response with whatever signatures it has, not checking if they are sufficient. So we don't need multiple `Item` implementations, actually.

Reply:
Anyway, the early return is probably a good idea and has been implemented here: https://github.com/casper-network/casper-node/pull/3125/commits/00edf29c255d56f93db9f6ce35aacdc944fbcd0f

@Fraser999 @goral09 
Let me know if you think it's correct to still update the metrics in such case.